### DISCUSSION TITLE
Best practices to prevent or reduce bugs?

### DISCUSSION THREAD

**Discussion by coiby** [OpenSanghaFoundation/OSF#54] [General]
The goal is to establish best practices for the OSF website development and maintenance.  So eventually we'll have the an efficient workflow to prevent/reduce bugs.

### Test a bug fix or feature before releasing it
1. Set up a local development environment for initial coding and testing.
2. Test the code in a staging site

### Require peer review for any significant change

For any change, require at least one person to review and approve the change. This helps
1. catch any issue beforehand
3. avoid stepping on other people's toes. People will be responsible for his/her area but quite often a change will bring unexpected effect.

### Use Git Git version control 

We may lose context of why a code was added in the first place. And we may easily make mistakes by forgetting the history and fix one issue at the cost of breaking one already working feature.

### Regularly update WordPress 

Keep WordPress, themes, and plugins updated to the latest versions to avoid attacker making use of known vulnerabilities..

### How shall we decide if we should guarantee WP administrator access to a person?

**Comment by nowmichaelclark** [comment]
For keeping wordpress, themes and plugins up to date; what would be a worthwhile way for doing this? Do we need a testing environment first where plugins are updated? Can we keep well trusted plug-ins set to auto-update?

**Reply by coiby** [reply]
Maybe we can turn auto-update on before officially launching OSF. Btw, do you know how to create a local testing environment?

**Reply by nowmichaelclark** [reply]
> do you know how to create a local testing environment?

No, I do not. It doesn't sound too difficult.

**Reply by coiby** [reply]
Great, then we can batch-update and batch-test the plugins regularly locally since Dharmmarata has experience of updating some plugins breaking the website. Later we consider using [Hostinger to create a testing env](https://www.hostinger.com/tutorials/wordpress-staging-environment).

**Comment by OSFOSF** [comment]
I do not know how to create a local testing environment.
what is auto-update.  there are working plugins that break when updated so i am not a fan of auto updating plugins.

the issue of autologin has been fixed,  If a user has never logged in befor we allow them to set the password, check it update it and then send them to the login page to login with it so that issue is fixed.
    On Monday, July 15, 2024 at 07:32:01 PM GMT+7, Coiby ***@***.***> wrote:  
 
 


Maybe we can turn auto-update on before officially launching OSF. Btw, do you know how to create a local testing environment?

**Comment by OSFOSF** [comment]
answers imbedded below
    On Sunday, July 14, 2024 at 09:55:22 AM GMT+7, Coiby ***@***.***> wrote:  
 
 


The goal is to establish best practices for the OSF website development and maintenance. So eventually we'll have the an efficient workflow to prevent/reduce bugs.

Some ideas

Test a bug fix or feature before releasing it
   
   - Set up a local development environment for initial coding and testing.
   - Test the code in a staging site
WE DO THAT ON LINE AND TEST ONLINE AND THEN FIX IT UNTILL IT WORKS ONLINE AS I HAVE NO CLUE TO SET UP A TEST BED

Require peer review for any significant change

For any change, require at least one person to review and approve the change. This helps
   
   - catch any issue beforehand
   - avoid stepping on other people's toes. People will be responsible for his/her area but quite often a change will bring unexpected effect.

WE DO THAT, I ask Micky to test after i do it.

Use Git Git version control

We may lose context of why a code was added in the first place. And we may easily make mistakes by forgetting the history and fix one issue at the cost of breaking one already working feature.


WE are doing that already, I do not forget the history of what I have done and the code is there to remind me.   Since I am the only one coding, there is no problem, we use github to report bugs we have been. 


Regularly update WordPress

Keep WordPress, themes, and plugins updated to the latest versions to avoid attacker making use of known vulnerabilities..

i DO NOT WANT TO DO THIS AS IT CREATES AN UNSTABLE ENVIRONMENT I do not trust all the plugins,  better to keep things working than up to date.

Questions

How shall we decide if we should guarantee WP administrator access to a person?

When someone has a skill we need that requires admin.  as of now there are too many admins and that can be cleaned out as most of them are not active
